### PR TITLE
Nova Hypervisors API

### PR DESCRIPTION
**Pull request by spjmurray**
Adds in support to list all available hypervisors (detailed) for applications
such as monitoring and capacity planning.

API: https://github.com/openstack/nova/blob/master/nova/api/openstack/compute/hypervisors.py#L46

Pay particular attention to line 71 where the horrible cpu_info hack is introduced

For #179

### CODE REVIEW

**Comment by jtopjian** [comment]
@spjmurray Nice! I'd love to see this added.

Can you open a corresponding Issue per the [style guide](https://github.com/gophercloud/gophercloud/blob/master/STYLEGUIDE.md)? This helps separate feature discussion vs code review discussion.

For the code itself, one question I have is whether or not this should go under directly under the `compute/v2` package or `compute/v2/extensions`. The previous API docs had this listed as an extension, but now everything is bundled under compute. @jrperritt: thoughts?

**Comment by spjmurray** [comment]
Sure thing @jtopjian. *Reads*  Also FYI the end goal of this is related to https://github.com/influxdata/telegraf/issues/243

**Comment by jtopjian** [comment]
@spjmurray This just might make me revisit InfluxDB now...

**Comment by spjmurray** [comment]
@jtopjian Here's some more fun InfluxDB related things I've done to really stoke a fire ;p http://www.datacentred.co.uk/blog/icinga-2-5-released/ http://www.datacentred.co.uk/blog/ceph-monitoring-telegraf-grafana/

**Comment by jrperritt** [comment]
@spjmurray is this still correctly labeled WIP or is it ready for review?

**Comment by spjmurray** [comment]
Apologies @jtopjian this got buried in my inbox and forgotten about.  Yes this is ready for review, I've stripped the WIP out of the pull request and the commit message.

File: openstack/compute/v2/hypervisors/results.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	Cores   int `json:"cores"`
+	Threads int `json:"threads"`
+}
+
+type CpuInfo struct {
+	Vendor   string   `json:"vendor"`
+	Arch     string   `json:"arch"`
+	Model    string   `json:"model"`
+	Features []string `json:"features"`
+	Topology `json:"topology"`
```
Comment (jrperritt) [inline openstack/compute/v2/hypervisors/results.go]: Let's name this field (`Topology`).

File: openstack/compute/v2/hypervisors/results.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+type Service struct {
+	Host           string `json:"host"`
+	ID             int    `json:"id"`
+	DisabledReason string `json:"disabled_reason"`
+}
+
+type Hypervisor struct {
+	// A structure that contains cpu information like arch, model, vendor, features and topology
+	CpuInfo `json:"cpu_info"`
```
Comment (jrperritt) [inline openstack/compute/v2/hypervisors/results.go]: Let's change this to:
```go
CPUInfo CPUInfo `json:"-"`
```

File: openstack/compute/v2/hypervisors/results.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	// The disk used in this hypervisor in GB
+	LocalGBUsed int `json:"local_gb_used"`
+	// The memory of this hypervisor in MB
+	MemoryMB int `json:"memory_mb"`
+	// The memory used in this hypervisor in MB
+	MemoryMBUsed int `json:"memory_mb_used"`
+	// The number of running vms on this hypervisor
+	RunningVMs int `json:"running_vms"`
+	// The hypervisor service object
+	Service `json:"service"`
```
Comment (jrperritt) [inline openstack/compute/v2/hypervisors/results.go]: Let's name this field (`Service`).

File: openstack/compute/v2/hypervisors/results.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	// The free disk remaining on this hypervisor in GB
+	FreeDiskGB int `json:"free_disk_gb"`
+	// The free RAM in this hypervisor in MB
+	FreeRamMB int `json:"free_ram_mb"`
+	// The hypervisor host name
+	HypervisorHostname string `json:"hypervisor_hostname"`
+	// The hypervisor type
+	HypervisorType string `json:"hypervisor_type"`
+	// The hypervisor version
+	HypervisorVersion int `json:"hypervisor_version"`
```
Comment (jrperritt) [inline openstack/compute/v2/hypervisors/results.go]: Let's change this to:
```go
HypervisorVersion int `json:"-"`
```

File: openstack/compute/v2/hypervisors/results.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+	if err != nil {
+		return err
+	}
+
+	*h = Hypervisor(hypervisor.tmp)
+
+	// Newer versions pass the CPU into around as the correct types, this just needs
+	// converting and copying into place. Older versions pass CPU info around as a string
+	// and can simply be unmarshalled by the json parser
+	switch t := hypervisor.CpuInfo.(type) {
```
Comment (jrperritt) [inline openstack/compute/v2/hypervisors/results.go]: Is there any reason this couldn't be simplified to this:
https://github.com/gophercloud/gophercloud/compare/master...jrperritt:compute_hypervisors_list#diff-cf3e97a140d10e9f8115943de0bab133R93

File: openstack/compute/v2/hypervisors/results.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+		h.HypervisorVersion = int(t)
+	default:
+		return fmt.Errorf("Hypervisor version of unexpected type")
+	}
+
+	return nil
+}
+
+type HypervisorPage struct {
+	pagination.LinkedPageBase
```
Comment (jrperritt) [inline openstack/compute/v2/hypervisors/results.go]: If this is a linked page, it should have a way to get from one page to the next (usually a `Links` field). Does that exist?
I noticed there are `Marker` and `Limit` fields in the `ListOpts`, so maybe it's a `MarkerPage`?

File: openstack/compute/v2/hypervisors/results.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+	RunningVMs int `json:"running_vms"`
+	// The hypervisor service object
+	Service `json:"service"`
+	// The number of vcpu in this hypervisor
+	VCPUs int `json:"vcpus"`
+	// The number of vcpu used in this hypervisor
+	VCPUsUsed int `json:"vcpus_used"`
+}
+
+func (h *Hypervisor) UnmarshalJSON(b []byte) error {
```
Comment (jrperritt) [inline openstack/compute/v2/hypervisors/results.go]: Have a look at the other `UnmarshalJSON` functions for style changes in this method (variable name changes).

File: openstack/compute/v2/hypervisors/results.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+	}
+
+	return nil
+}
+
+type HypervisorPage struct {
+	pagination.LinkedPageBase
+}
+
+func (page HypervisorPage) IsEmpty() (bool, error) {
```
Comment (jrperritt) [inline openstack/compute/v2/hypervisors/results.go]: Have a look at the other UnmarshalJSON functions for style changes in this method (variable name changes).

File: openstack/compute/v2/hypervisors/results.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+	hypervisors, err := ExtractHypervisors(page)
+	return len(hypervisors) == 0, err
+}
+
+func ExtractHypervisors(p pagination.Page) ([]Hypervisor, error) {
```
Comment (jrperritt) [inline openstack/compute/v2/hypervisors/results.go]: Have a look at the other UnmarshalJSON functions for style changes in this method (variable name changes).

File: openstack/compute/v2/hypervisors/results.go (hunk 6 above)
Comment (spjmurray) [inline openstack/compute/v2/hypervisors/results.go]: @jrperritt You are right, it's actually a marker page.  Sadly when I implemented it properly it seems our version of OpenStack doesn't support the marker field (e.g. needs Nova V2.33) and just spins forever.  Do we have a standard way of catering for OpenStack's API stupidity regarding backwards compatibility?

Comment (spjmurray) [inline openstack/compute/v2/hypervisors/results.go]: I'm guessing lowest common denominator e.g. SinglePageBase

File: openstack/compute/v2/hypervisors/results.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+	// The number of vcpu in this hypervisor
+	VCPUs int `json:"vcpus"`
+	// The number of vcpu used in this hypervisor
+	VCPUsUsed int `json:"vcpus_used"`
+}
+
+func (r *Hypervisor) UnmarshalJSON(b []byte) error {
+
+	type tmp Hypervisor
+	var s *struct {
```
Comment (jtopjian) [inline openstack/compute/v2/hypervisors/results.go]: @jrperritt I'm unable to find Go documentation to explain when `*struct` is required. All other `UnmarshalJSON` functions use `struct`. Testing with both seems to result in the same.

Is `*struct` required here?

Comment (jtopjian) [inline openstack/compute/v2/hypervisors/results.go]: To clarify a little bit:

I can do some separate / local tests with `struct` and `*struct` and see the differences there, but it seems here, `json.Unmarshal` will happily take both. Just not sure if `*struct` should be respected here.

Comment (jrperritt) [inline openstack/compute/v2/hypervisors/results.go]: > Is *struct required here?

No, and we should maintain consistency with the other `UnmarshalJSON` methods [by changing it to a non-pointer].

**Comment by jtopjian** [comment]
I've cleaned this up a bit and going to resubmit this in a new PR. @spjmurray's changes will remain in the commit history.

**Pull request closed without merging**